VIPs/vip-10.md

```diff
@@ -0,0 +1,109 @@
+---
+vip: 10
+title: Storage node registration (on-chain)
+author: Chris Were <chris@verida.io>
+discussions-to: #dev-protocol on Verida Discord
+status: Draft
+type: Standards Track
+category: Core
+created: 2023-03-22
+---
+
+# Overview
+
+The selection of storage nodes occurs when a DID is first created and when a context is connected to for the first time. The storage nodes are used to store the actual DID Documents and the same nodes are typically (but not necessarily) used to store data for each application context.
+
+This VIP provides a specification for storing the available storage nodes on-chain.
+
+This is phase 2 of [VIP-7: Storage node discovery and selection (off-chain)](./vip-7.md)
+
+This VIP does not implement any token based staking or slashing of nodes, it's purely focused on node addition, removal and discovery.
+
+# Motivation
+
+The current storage node database is centralized and needs to be stored on-chain to decentralize the discovery and selection of nodes on the network.
+
+# Specification
+
+## Data structures
+
+### StorageNode
+
+- didAddress: string
+- endpointUri: string
+- countryCode: string
+- regionCode: string
+- datacenterId: id
+
+### Datacenter
+
+- id: int
+- countryCode: string
+- regionCode: string
+- lat: int
+- long: int
+
+## State variables
+
+TBA
+
+## addNode(didAddress: string, endpointUri: string, countryCode: string, regionCode: string, datacenterId: string, requestSignature: string)
+
+Storage nodes must register themselves with the protocol so they can be discovered by new users. Storage nodes provide additional metadata (via `/status` endpoint) that assists users select the most appropriate storage nodes to use.
+
+It's not possible to re-use a DID to register multiple storage nodes. This is because storage endpoints sign their responses using the DID private key to enhance security of the protocol.
+
+This method registers a new endpoint on the network:
+
+1. `didAddress`: The DID that is associated with the storage node. Must be only one `didAddress` registered at any time.
+2. `endpointURI`: The storage node endpoint. Must be only one `endpointURI` registered at any time.
+3. `countryCode`: Unique two-character string code (see [VIP-7](./vip-7.md))
+4. `regionCode`: Unique region string code (see [VIP-7](./vip-7.md))
+5. `datacenterId`: Unique datacenter identifier. Must match a datacenter previously created with `addDatacenter()`.
```

Comment:
No, this is intentional.

At this stage, only we should be able to add data centres to avoid duplications etc.

In the future there could be a more complex decentralized voting process to add that type of info.

Reply:
Yep, I get the `ownerOnly` for creating a datacenter.
I was pointing the "previously created" may suggest they (the storage nodes) have to create the datacenter, but they should not (they can't), they must use an existing one.

Reply:
Got it. Adjusted wording.